core/src/main/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.scala

```diff
@@ -0,0 +1,132 @@
+      val fileBufferSize = conf.getSizeAsKb("spark.shuffle.file.buffer", "32k").toInt * 1024
+      val openStartTime = System.nanoTime
+      partitionWriters = Array.fill(numPartitions) {
+        val (blockId, file) = blockManager.diskBlockManager.createTempShuffleBlock()
+        val writer =
+          blockManager.getDiskWriter(blockId, file, serInstance, fileBufferSize, writeMetrics)
+        writer.open()
+      }
+      // Creating the file to write to and creating a disk writer both involve interacting with
+      // the disk, and can take a long time in aggregate when we open many files, so should be
+      // included in the shuffle write time.
+      writeMetrics.incShuffleWriteTime(System.nanoTime - openStartTime)
+
+      while (records.hasNext) {
+        val record = records.next()
+        val key: K = record._1
+        partitionWriters(partitioner.getPartition(key)).write(key, record._2)
+      }
+    }
+  }
+
+  /**
+   * Write all the data added into this writer into a single file in the disk store. This is
+   * called by the SortShuffleWriter and can go through an efficient path of just concatenating
+   * the per-partition binary files.
+   *
+   * @param blockId block ID to write to. The index file will be blockId.name + ".index".
+   * @param context a TaskContext for a running Spark task, for us to update shuffle metrics.
+   * @return array of lengths, in bytes, of each partition of the file (used by map output tracker)
+   */
+  def writePartitionedFile(blockId: BlockId, context: TaskContext, file: File): Array[Long] = {
+    if (partitionWriters == null) {
+      // We were passed an empty iterator
+      Array.fill(numPartitions)(0L)
+    } else {
+      partitionWriters.foreach(_.commitAndClose())
+
+      // Track location of each range in the output file
+      val lengths = new Array[Long](numPartitions)
+
+      val transferToEnabled = conf.getBoolean("spark.file.transferTo", true)
+      val out = new FileOutputStream(file, true)
+      val writeStartTime = System.nanoTime
+      Utils.tryWithSafeFinally {
+        for (i <- 0 until numPartitions) {
+          val in = new FileInputStream(partitionWriters(i).fileSegment().file)
+          Utils.tryWithSafeFinally {
+            lengths(i) = Utils.copyStream(in, out, closeStreams = false, transferToEnabled)
+          } {
+            in.close()
+          }
+          if (!blockManager.diskBlockManager.getFile(partitionWriters(i).blockId).delete()) {
+            logError("Unable to delete file for partition i. ")
+          }
+        }
+      } {
+        out.close()
+        context.taskMetrics().shuffleWriteMetrics.foreach { m =>
+          m.incShuffleWriteTime(System.nanoTime - writeStartTime)
+        }
+      }
+
+      lengths
+    }
+  }
+
+  def stop(): Unit = {
+    if (partitionWriters != null) {
+      partitionWriters.foreach { w =>
+        w.revertPartialWritesAndClose()
```

Comment:
Actually, this method explicitly doesn't throw exceptions.  From its Scaladoc:

```
  /**
   * Reverts writes that haven't been flushed yet. Callers should invoke this function
   * when there are runtime exceptions. This method will not throw, though it may be
   * unsuccessful in truncating written data.
   */
  def revertPartialWritesAndClose()
```